### ISSUE TITLE
[Unexpected Behavior] Game framerate seems to affects polling rate

### ISSUE THREAD

**Issue by 48productions** [TeamRizu/OutFox#48]
**What you're using**

On the latest version of Stepmania 5.3, limiting the game's framerate using VSync in-game or with external tools seems to introduce artifacts in the timing charts on the Simply Love/Soundwaves themes that suggest input polling issues (distinct lines and spikes in the offset graphs, screenshots linked below under each person involved in this).

This issue seems to be mostly hardware/OS independent (as I've confirmed this issue on Win 10/NVidia and Debian/AMD, among other setups other people have reported this on), and seems to not appear on Stepmania 5.1 (also confirmed by myself and others).

In addition to [my findings](https://cdn.discordapp.com/attachments/428678897815781378/733856746237329448/Help_VSync_is_Plaguing_my_Life.pdn), credit to [shockdude](https://cdn.discordapp.com/attachments/428678897815781378/734490557727899658/unknown.png), [Haley Halcyon](https://cdn.discordapp.com/attachments/428678897815781378/738819821789118564/unknown.png), and [Mr.](https://cdn.discordapp.com/attachments/428678897815781378/738831602427101344/unknown.png) [ThatKid](https://cdn.discordapp.com/attachments/428678897815781378/738827317710946375/2020-07-31_113525.jpg) for also confirming this issue on different setups/songs/themes, and [slaugaus](https://ptb.discordapp.com/channels/422897054386225173/428678897815781378/733473575012663416) for the initial report with my screenshots (see #troubleshooting on the Project Moondance discord for more context/information).

Also thanks to Yumry/Crazy4321/stevereen for helping identify this issue on other discord servers as well.

**Expected behavior**

Input seems like it shouldn't be tied to the game's framerate, since you'd be able to arbitrarily improve your timing by turning off vsync.

**Comment by shockdude** [comment]
Thanks for the credit! Looking forward to a fix.
Other screenshot of mine attached.
![image](https://user-images.githubusercontent.com/27760396/89088989-e4a33600-d34f-11ea-94c7-7fcf5d66e29d.png)

**Comment by Scraticus** [comment]
Right, so I've been investigating this, and i can confirm the following:

1. Input is not tied to frame-rate. if it was, you would have VERY different spikes.
2. di.ts (Device Input Time Stamp) does not report polling rate times, nor locked times when you're playing. I've confirmed this with multiple tests and with many different devices.
3. The _graph_ seems to be doing something weird here, as it shows a 'jagged' polling error but this is NOT actually the case. We need to investigate this further.
4. The Histogram needs to be also investigated. This is likely a maths problem over anything else here.
5. The Input Thread runs outside the 'vblank' loop, which was something i thought it was within, however, when vsync is on, we can report sub 16.67ms button presses. 
6. This is a massive rabbit hole and will take more time to identify what is at fault. The game however is NOT tied to frame-rates i can guarantee this.

Will update more when I have finished my investigation.

**Comment by Scraticus** [comment]
@shockdude @48productions can you run a test on the latest alpha for this please?

Many thanks

**Comment by shockdude** [comment]
Issue still exists in latest alpha.
![image](https://user-images.githubusercontent.com/27760396/90318488-a6784b80-dee5-11ea-9959-c119ac04f551.png)

**Comment by shockdude** [comment]
Tried again, Peak position seems inconsistent?
![image](https://user-images.githubusercontent.com/27760396/90318723-42ef1d80-dee7-11ea-83f3-aea460d844ba.png)
[Logs.zip](https://github.com/TeamRizu/OutFox/files/5079241/Logs.zip)

**Comment by 48productions** [comment]
Not totally sure how helpful it is/isn't at this point but I was able to mostly confirm this issue on the latest alpha as well. Same PC as before, but with different code for the pad (that decided to go full cursed mode for the first few songs before I enabled debounce in SM), and a 120BPM song instead of 180BPM. Not exactly the most scientific way of testing this, but it got the job done.

Worth nothing both joystick/gamepad scores were in the fancy new legacy mapping mode, as Stepmania hung when I tried to map the pad in the xinput mapping mode. I might play around with this a bit more in the future to see if I can get any more details on what specifically is going on.

![VSync did a thing MK2](https://user-images.githubusercontent.com/57228987/90324002-903eaf80-df26-11ea-97e3-6b8d955e6f58.png)

**Comment by Scraticus** [comment]
So after spending some more time on this, we have deduced the following.

- The thread is actually a Lua issue, there is a pass-through of (DeltaTime) causing an odd update to 60 FPS locking.
- The _scaling_ of the graph affects the stepping.
- We are looking into how to display this better when the points are 'lumped together' or 'stepped'.
- Input is definitely on it's own thread, - there are no issues there.

We'll keep on investigating. This is still on our 'wtf is going on' list.

**Comment by Scraticus** [comment]
Right, this has been identified and finally squashed thanks to lots of head scratching from me and Jous to get the solution.

The issue is due to the way we initialised the Low Level window, and the way Input was initialised. in the TL;DR of it, The loop for input could slip into using the message loops `fDeltaTime` for the timestamps (which were aligned to Vsync).

It didn't happen all the time, and it was weird for us here as it was tough to isolate. We have fixed it by moving _everything_ input uses to its own thread, so there is no possible chance for times to weird out. The way we did the threading initialisation was not as optimised as we could, so it has been rewritten to be better. We have fixed a few issues with keyboards and some legacy joysticks using this system. It'll need some testing yet, but I am pleased to say it's behind us.

Thanks for the patience on this one, it's been a pretty crazy ride!

**Comment by shockdude** [comment]
Is this still being worked on? The bug is still present in 4.9.6
![alpha496issue48](https://user-images.githubusercontent.com/27760396/105562306-82a80580-5cce-11eb-925b-a0f626c0b400.png)
